src/component/1d/matrix/Stocsy.tsx

```diff
@@ -0,0 +1,159 @@
+import { extent } from 'd3';
+import { xFindClosestIndex } from 'ml-spectra-processing';
+import { matrixToStocsy } from 'nmr-processing';
+import { useMemo } from 'react';
+
+import { useChartData } from '../../context/ChartContext';
+import { useScaleChecked } from '../../context/ScaleContext';
+import { usePanelPreferences } from '../../hooks/usePanelPreferences';
+import { PathBuilder } from '../../utility/PathBuilder';
+import { getYScaleWithRation } from '../utilities/scale';
+
+import {
+  findXFromToIndex,
+  interpolatedColorsPoints,
+  sliceArray,
+  useMatrix,
+} from './useMatrix';
+
+interface StocsyProps {
+  x: Float64Array | number[];
+  y: number[];
+  color: string[];
+  scaleRatio: number;
+  yDomain: number[];
+}
+
+interface StocsyIndexPointProps {
+  chemicalShift: number | null;
+}
+
+function useStocsy(chemicalShift: number) {
+  const matrix = useMatrix();
+  const {
+    xDomain: [from, to],
+  } = useChartData();
+
+  return useMemo(() => {
+    if (!matrix) return null;
+
+    const { x, matrixY } = matrix;
+
+    const cIndex = xFindClosestIndex(x, chemicalShift ?? x[0]);
+    const { color, y } = matrixToStocsy(matrixY, cIndex);
```

Comment:
matrixToStocsy should not be called when we zoom and I don't think the Memo should be triggered by from and to.